tls/resource_cert_revocation_list.go

```diff
@@ -0,0 +1,193 @@
+			Sensitive:   true,
+			StateFunc: func(v interface{}) string {
+				return hashForState(v.(string))
+			},
+		},
+		"ca_key_algorithm": &schema.Schema{
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Name of the algorithm used to generate the certificate's private key",
+			ForceNew:    true,
+		},
+	}
+
+	return &schema.Resource{
+		Create:        CreateCRL,
+		Delete:        DeleteCRL,
+		Read:          ReadCRL,
+		Update:        UpdateCRL,
+		CustomizeDiff: CustomizeCertificateDiff,
+		Schema:        s,
+	}
+}
+
+func CreateCRL(d *schema.ResourceData, meta interface{}) error {
+	certsToRevoke := make([]pkix.RevokedCertificate, 0)
+	if value := d.Get("certs_to_revoke").([]interface{}); len(value) > 0 {
+		for _, vi := range value {
+			block, err := decodePEMFromBytes([]byte(vi.(string)), "")
+			if err != nil {
+				return err
+			}
+			certificate, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return err
+			}
+			certsToRevoke = append(certsToRevoke, pkix.RevokedCertificate{
+				SerialNumber: certificate.SerialNumber,
+			})
+		}
+	}
+	caKey, err := parsePrivateKey(d, "ca_private_key_pem", "ca_key_algorithm")
+	if err != nil {
+		return err
+	}
+	caCert, err := parseCertificate(d, "ca_cert_pem")
+	if err != nil {
+		return err
+	}
+
+	notBefore := now()
+	notAfter := notBefore.Add(time.Duration(d.Get("validity_period_hours").(int)) * time.Hour)
+	validFromBytes, err := notBefore.MarshalText()
+	if err != nil {
+		return err
+	}
+	validToBytes, err := notAfter.MarshalText()
+	if err != nil {
+		return err
+	}
+
+	crlBytes, err := caCert.CreateCRL(rand.Reader, caKey, certsToRevoke, notBefore, notAfter)
+	if err != nil {
+		return err
+	}
+
+	crlPem := string(pem.EncodeToMemory(&pem.Block{Type: "X509 CRL", Bytes: crlBytes}))
+	if err != nil {
+		return err
+	}
+	d.SetId(hashForState(string(crlBytes)))
+	d.Set("crl_pem", crlPem)
+	d.Set("ready_for_renewal", false)
+	d.Set("validity_start_time", string(validFromBytes))
+	d.Set("validity_end_time", string(validToBytes))
+	return nil
+}
+
+func DeleteCRL(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("")
+	return nil
+}
+
+func ReadCRL(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func UpdateCRL(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func CustomizeCrlDiff(d *schema.ResourceDiff, meta interface{}) error {
+	var readyForRenewal bool
+
+	endTimeStr := d.Get("validity_end_time").(string)
+	endTime := now()
+	err := endTime.UnmarshalText([]byte(endTimeStr))
+	if err != nil {
+		// If end time is invalid then we'll treat it as being at the time for renewal.
+		readyForRenewal = true
+	} else {
+		earlyRenewalPeriod := time.Duration(-d.Get("early_renewal_hours").(int)) * time.Hour
+		endTime = endTime.Add(earlyRenewalPeriod)
+
+		currentTime := now()
+		timeToRenewal := endTime.Sub(currentTime)
+		if timeToRenewal <= 0 {
+			readyForRenewal = true
+		}
+	}
+
+	if readyForRenewal {
```

Comment:
I would move this block of code to separate function, let's call it `markForRenewal()`, so then at the beginning of this function, you could do the following:
```go
	err := endTime.UnmarshalText([]byte(endTimeStr))
	if err != nil {
          return markForRenewal(d)
        }
```

This way, you can avoid using `else`, which generally should be avoided in Go (e.g. https://medium.com/@jamousgeorges/dont-use-else-66ee163deac3).

`markForRenewal()` would be something like:
```go
func markForRenewal(d *schema.ResourceDiff) error {
  err := d.SetNew("ready_for_renewal", true)
  if err != nil {
    return err
  }

  return d.ForceNew("ready_for_renewal")
}
```

Reply:
Done, love this suggestion! it reduces the cognitive complexity making it way easier to understand, thanks! Thanks also for the article reference, interesting! :)